### PR TITLE
Genesis block

### LINKED ISSUE #1032: Genesis builder
Information for genesis block has to come from https://github.com/witnet/genesis_block/blob/master/genesis_block.json, has no mint reward and only will be mine for nodes that has a specific flag for it.

Genesis block hash and bootstrap hash has to be added in config file

### LINKED ISSUE #1036: Genesis validation
A special validation for the geneisis block is needed

### PR DESCRIPTION
**Pull request by tmpolaczyk**
Close #1032
Close #1036

Now:

* nodes can sync regardless if they choose to create the genesis block or not
* the UTXOs from the genesis block can be spent
* genesis block validation just returns true
* an example genesis block is available at https://github.com/witnet/genesis_block/

Remaining issues:
* verify that timelock is respected
* verify that sheikah can spend UTXOs from the genesis block
* should we stop the node if genesis_mining = true and genesis block creation (or consolidation) fails? (it currently panics)
* is it possible to sync before epoch 0?
* make sure that nodes with bootstrap_hash as the beacon never try to mine normal blocks
* add documentation

### CODE REVIEW

File: node/src/actors/chain_manager/actor.rs [hunk 1]
```diff
@@ -133,24 +154,25 @@ impl ChainManager {
                         if x.is_some() {
                             debug!(
                                 "Uninitialized local chain the ChainInfo in storage is incomplete. Proceeding to \
-                                 initialize and store a new chain."
+                                     initialize and store a new chain."
```
Comment (tmpolaczyk) [inline node/src/actors/chain_manager/actor.rs]: fmt

File: data_structures/src/serialization_helpers.rs [hunk 2]
```diff
@@ -6,9 +6,13 @@
...
 // Ideally all this code would be generated with a `#[serde(human_readable_string)]` macro.
 
-use crate::chain::{Hash, OutputPointer, PublicKeyHash, SHA256};
+use crate::chain::{
+    GenesisBlockInfo, Hash, OutputPointer, PublicKeyHash, ValueTransferOutput, SHA256,
+};
 use crate::get_environment;
 use serde::{de, Deserialize, Deserializer, Serialize, Serializer};
+use std::fmt::Display;
+use std::str::FromStr;
```
Comment (lrubiorod) [inline data_structures/src/serialization_helpers.rs]: fmt

File: validations/src/validations.rs [hunk 3]
```diff
@@ -1280,6 +1340,53 @@ pub fn validate_block(
...
+
+    let merkle_roots = BlockMerkleRoots {
+        mint_hash: txns.mint.hash(),
+        vt_hash_merkle_root: merkle_tree_root(&txns.value_transfer_txns),
+        dr_hash_merkle_root: merkle_tree_root(&txns.data_request_txns),
+        commit_hash_merkle_root: merkle_tree_root(&txns.commit_txns),
+        reveal_hash_merkle_root: merkle_tree_root(&txns.reveal_txns),
+        tally_hash_merkle_root: merkle_tree_root(&txns.tally_txns),
+    };
+    let empty_block = Block {
```
Comment (lrubiorod) [inline validations/src/validations.rs]: it is not an empty block, should we use another name?

**Comment by lrubiorod** [comment]
When you use the cli command blockchain, genesis block doesnt appear

File: validations/src/validations.rs (hunk 3 above)
Comment (tmpolaczyk) [reply validations/src/validations.rs]: `block_with_most_fields_set_to_the_default_value` doesn't sound quite good to me :laughing:

**Comment by tmpolaczyk** [comment]
> When you use the cli command blockchain, genesis block doesnt appear

True, that looks like a bug in the CLI command, as it works if you manually set the epoch to 0:

```
$ cargo run -- -c witnet_01.toml node blockchain --epoch=0 --limit=100
    Finished dev [unoptimized + debuginfo] target(s) in 0.15s
     Running `target/debug/witnet -c witnet_01.toml node blockchain --epoch=0 --limit=100`
Loading config from: witnet_01.toml
Setting log level to: ERROR, source: Config
block for epoch #0 had digest 9de38ff00b532f0fd52ec2c063e4de4a414f820e6c1ef0f2bb6e815c18459f7e
$ cargo run -- -c witnet_01.toml node blockchain
    Finished dev [unoptimized + debuginfo] target(s) in 0.15s
     Running `target/debug/witnet -c witnet_01.toml node blockchain`
Loading config from: witnet_01.toml
Setting log level to: ERROR, source: Config
```

**Comment by aesedepece** [comment]
I support @tmpolaczyk's suggestion. Let's have an issue for that.

File: node/src/actors/chain_manager/actor.rs [hunk 4]
```diff
@@ -73,9 +73,25 @@ impl ChainManager {
...
-                // Store the genesis block hash
+                // Store the bootstrap hash, genesis block hash and genesis mining flag
+                act.bootstrap_hash = config.consensus_constants.bootstrap_hash;
                 act.genesis_block_hash = config.consensus_constants.genesis_hash;
 
+                let info_genesis = act.info_genesis.take().or_else(|| if config.mining.genesis_mining {
+                    Some(
+                        GenesisBlockInfo::from_path(&config.mining.genesis_path, act.bootstrap_hash, act.genesis_block_hash)
+                            .map_err(|e| log::error!("Failed to create genesis block: {}", e))
+                            // TODO: gracefully stop node
```
Comment (tmpolaczyk) [inline node/src/actors/chain_manager/actor.rs]: Maybe with:

    System::current().stop_with_code(1);

**Pull request merged**